### ISSUE TITLE
Annotation processing is stuck in eclipse

### ISSUE THREAD

**Issue by WonderCsabo** [johncarl81/parceler#30]
I use eclipse to develop and build my projects. From the command line, i can compile with maven, but unfortunately eclipse cannot do that. The annotation processing is started, but somewhere it is stuck. I debugged the processing, and i made it to `TransactionProcessorPool.execute()`. Here the `ExecutorService.execute()` method is called, but the execution is stuck at `ExecutorService.awaitTermination()`. I guess the `Transaction` goes into a forever loop or something. Unfortunately i could not debug it any deeper.

I am using eclipse 4.3.2 on Windows 8. [This](https://drive.google.com/file/d/0Bx0_b672V3qnOGQzSllEN3pZZjA/edit?usp=sharing) is the sample project what i used.

**Comment by titanseason** [comment]
Hi, I find a project at https://github.com/baoyongzhang/ParcelableGenerator, using eclipse to generate Parcelable.

**Comment by WonderCsabo** [comment]
I debbuged a little bit again, and it seems to threads are waiting on the same object. I think Parceler creates a deadlock in JDT APT. I try to look into it later.

**Comment by johncarl81** [comment]
@WonderCsabo, I appreciate you looking.  My guess is that the transactional approach that is used is messing with how Eclipse expects a processor to work.  We do use a single thread executor, but it does use a thread off of the original processor:
https://github.com/johncarl81/transfuse/blob/master/transfuse-support/src/main/java/org/androidtransfuse/transaction/TransactionProcessorPool.java#L56
I guess we could try to use the original thread here to try and avoid deadlock.

@titanseason, there are a number of solutions out there.  I think Parceler is the most complete and user friendly (except for the Eclipse integration I'm sorry to say).  Ugh, looks like ParcelableGenerator uses reflection heavily behind the annotation processor and doesn't support Collections.  I can't say I approve.

**Comment by WonderCsabo** [comment]
@johncarl81 can you create an test Parceler which uses on thread? If you do so, i can check it in Eclipse.

**Comment by johncarl81** [comment]
Sure, I'll get something to you this afternoon.

**Comment by johncarl81** [comment]
Here's the branch: https://github.com/johncarl81/parceler/tree/single_thread  Do you need an artifact?

**Comment by WonderCsabo** [comment]
mvn install should do it, shouldn't it?

**Comment by WonderCsabo** [comment]
Also, i just want to confirm the the branch only contains a change in the
POM. Is that correct?

**Comment by johncarl81** [comment]
yes, that's correct... I just updated the transfuse dependency.

I just deployed the artifact if that's more convenient: https://oss.sonatype.org/content/repositories/snapshots/org/parceler/parceler/1.0.2-SNAPSHOT/parceler-1.0.2-20150626.204834-2.jar

**Comment by WonderCsabo** [comment]
Unfortunetaly it still stuck at:
```
Thread [pool-4-thread-1] (Suspended)	
	owns: OutputStreamWriter  (id=566)	
	owns: CodeWriter$1  (id=567)	
	owns: BufferedWriter  (id=568)	
	waiting for: Object  (id=569)	
	Object.wait(long) line: not available [native method]	
	Object.wait() line: 502	
	ThreadJob.waitForRun(ThreadJob, IProgressMonitor, InternalJob, Thread) line: 270	
	ThreadJob.joinRun(ThreadJob, IProgressMonitor) line: 197	
	ImplicitJobs.begin(ISchedulingRule, IProgressMonitor, boolean) line: 92	
	JobManager.beginRule(ISchedulingRule, IProgressMonitor) line: 307	
	WorkManager.checkIn(ISchedulingRule, IProgressMonitor) line: 120	
	Workspace.prepareOperation(ISchedulingRule, IProgressMonitor) line: 2189	
	Workspace.run(IWorkspaceRunnable, ISchedulingRule, int, IProgressMonitor) line: 2236	
	CreatePackageFragmentOperation(JavaModelOperation).runOperation(IProgressMonitor) line: 794	
	PackageFragmentRoot.createPackageFragment(String, boolean, IProgressMonitor) line: 302	
	CompilationUnitHelper.createPackageFragment(String, IPackageFragmentRoot, IProgressMonitor) line: 144	
	GeneratedFileManager.generateFileDuringBuild(Collection<IFile>, String, String, boolean, IProgressMonitor) line: 530	
	IdeJavaSourceOutputStream.close() line: 67	
	StreamEncoder.implClose() line: 320	
	StreamEncoder.close() line: 149	
	OutputStreamWriter.close() line: 233	
	CodeWriter$1(FilterWriter).close() line: 104	
	BufferedWriter.close() line: 266	
	PrintWriter.close() line: 339	
	JFormatter.close() line: 137	
	JPackage.build(CodeWriter, CodeWriter) line: 439	
	JCodeModel.build(CodeWriter, CodeWriter) line: 311	
	CodeGenerationScopedTransactionWorker.innerRun(V) line: 49	
	CodeGenerationScopedTransactionWorker(AbstractCompletionTransactionWorker).run(V) line: 35	
	ScopedTransactionWorker.run(V) line: 55	
	Transaction.run() line: 77	
	ThreadPoolExecutor.runWorker(ThreadPoolExecutor$Worker) line: 1142	
	ThreadPoolExecutor$Worker.run() line: 617	
	Thread.run() line: 745	
```

**Comment by WonderCsabo** [comment]
OOps, i used the old processor, testing the new one.

**Comment by WonderCsabo** [comment]
@johncarl81 seems to be working! Where i can find test projects the test all situations?

**Comment by johncarl81** [comment]
@WonderCsabo, Great news!  This will be covered extensively by building transfuse and parceler, which are passing ATM:

https://travis-ci.org/johncarl81/parceler/builds/68529594
https://travis-ci.org/johncarl81/transfuse/builds/68528990

If you want you can build the branch I referenced and build the example projects.  That should exercise this functionality to the n-th degree.

FYI: https://github.com/johncarl81/transfuse/pull/177

Let me know if you need anything else.

**Comment by WonderCsabo** [comment]
Great! I will import and build the example projects in Eclipse. However I can only work on next week on this.

**Comment by johncarl81** [comment]
No problem @WonderCsabo, any further input is much appreciated.  In the mean time I'll try to get a PR put together in our other effort.

Thanks again for the help on this.  It's great to work together!

**Comment by WonderCsabo** [comment]
@johncarl81 unfortunetaly i could not test the examples because of two problems with `@AutoValue`:

* AutoValue processor is crashing in Eclipse:
```
@AutoValue processor threw an exception: java.lang.NullPointerException
	at com.google.auto.value.processor.AutoValueProcessor.getTypeMirror(AutoValueProcessor.java:644)
	at com.google.auto.value.processor.AutoValueProcessor.defineVarsForType(AutoValueProcessor.java:406)
	at com.google.auto.value.processor.AutoValueProcessor.processType(AutoValueProcessor.java:390)
	at com.google.auto.value.processor.AutoValueProcessor.process(AutoValueProcessor.java:128)
```

* `javax.annotation.Generated` is not available on the classpath of Eclipse Android projects, hence the generated `@AutoValue` classess will have compilation errors.

BTW, i think you should not use `@AutoValue` in your example projects or other framework, you should focus only on your annotations. You may create an additional example which will use integrations, though.

**Comment by johncarl81** [comment]
I went ahead and removed AutoValue from the examples on master.  Go ahead and try again.

**Comment by WonderCsabo** [comment]
Great! I built and deployed the example app from Eclipse, it works. One annoying thing thought that it will mark references like `ExampleParcel$$Parcelable` as an error in the editor, but not in the problems list, and let's building and deploying the app.

**Comment by johncarl81** [comment]
Fantastic.

I'm not surprised the generated code shows up as an error.  Have you seen Eclipse pick up the generated classes before?  Should we be doing something different in Parceler?

**Comment by WonderCsabo** [comment]
Eclipse can pickup the generated classes (we always use that we AA too) ,
actually it finds them with ctrl + K. I think the problem is with the
dollar notation.

**Comment by johncarl81** [comment]
Strange.  Well, I always try to keep the developer from referencing generated code.  You should be using the `Parcels` utility to wrap and unwrap the annotated POJOs.

**Comment by WonderCsabo** [comment]
I know that. ;) That reference is in the example project, but that is OK.

**Comment by johncarl81** [comment]
I have no idea on this one... have you been able to build the Gradle project in Eclipse?

**Comment by WonderCsabo** [comment]
Sorry, but i am not sure what do you mean. Android Gradle projects are not working in Eclipse BTW, and simple Gradle projects just got support in the latest Mars release.

But i think there is no problem here... Users should not reference the generated code, and if they do,they just have to ignore the red marks in the editor.

**Comment by johncarl81** [comment]
Fair enough.

Thanks again for your efforts here @WonderCsabo, I really appreciate it! :beers:

**Comment by WonderCsabo** [comment]
BTW, i just tried to run the tests with the Eclipse JUnit 4 launcher and i got an NPE in `findSuperInterfaces`. I guess Eclipse is really not designed for this.

**Comment by WonderCsabo** [comment]
BTW, i found a bigger problem: Parceler does not really support incremental compilation, I just added a new class with `@Parcel` annotation, and it was added to the generated `Parceler$$Parcels` class, however the other, already present parcelable class disappeared...

**Comment by johncarl81** [comment]
I've been wondering about partial compilation, and I'm surprised it hasn't come up sooner.  It seems with a proper build in place partial compilation is a rarity these days.  That being said I think there are ways that we can handle partial compilation.

I wonder if the findSuperInterfaces is throwing an NPE also because the target type isnt being compiled (because of partial compilation).

**Comment by WonderCsabo** [comment]
I can debug findSuperInterfaces tomorrow, but incremental compilation is more interesting. AFAIK both Android Gradle (Jack and Jill) and Maven (Takari life cycle) teams are planning to rely on incremental compiler heavily.

**Comment by WonderCsabo** [comment]
I think we should open a separate issue for incremental.

**Comment by johncarl81** [comment]
Which means this is a good time to tackle it.

Agreed.

**Comment by WonderCsabo** [comment]
Let's open a separate issue, then. BTW, Parceler can still work if some classes are missing from the generated `Parcels` class, does it?

**Comment by WonderCsabo** [comment]
`findSuperInterfaces` is no longer a problem in Eclipse Mars, and all 44 tests passed in Eclipse.

**Comment by johncarl81** [comment]
Yes, the `Parceler$$Parcels` generated class is just a dictionary for the `Parcels` utility to map generated classes to the original annotated classes.  If the map does not contain an entry for the given annotated type then the `Parcels` utility will resort to reflection to lookup the generated class.

Thats good news...  The changes we just made to Transfuse (in particular this one: https://github.com/johncarl81/transfuse/pull/176) must have avoid the `findSuperInterfaces` call in question.

**Comment by WonderCsabo** [comment]
Should we open a separate issue for partial compilation?

**Comment by johncarl81** [comment]
Yes!
